lib_eio_posix/sched.ml

```diff
@@ -0,0 +1,348 @@
+
+  (* If [false], the main thread will check [run_q] before sleeping again
+     (possibly because an event has been or will be sent to [eventfd]).
+     It can therefore be set to [false] in either of these cases:
+     - By the receiving thread because it will check [run_q] before sleeping, or
+     - By the sending thread because it will signal the main thread later *)
+  need_wakeup : bool Atomic.t;
+
+  sleep_q: Zzz.t;                       (* Fibers waiting for timers. *)
+}
+
+(* The message to send to [eventfd] (any character would do). *)
+let wake_buffer = Bytes.of_string "!"
+
+(* This can be called from any systhread (including ones not running Eio),
+   and also from signal handlers or GC finalizers. It must not take any locks. *)
+let wakeup t =
+  Atomic.set t.need_wakeup false; (* [t] will check [run_q] after getting the event below *)
+  Rcfd.use t.eventfd
+    ~if_closed:ignore       (* Domain has shut down (presumably after handling the event) *)
+    (fun fd ->
+       (* This can fail if the pipe is full, but then a wake up is pending anyway. *)
+       ignore (Unix.single_write fd wake_buffer 0 1 : int);
+    )
+
+(* Safe to call from anywhere (other systhreads, domains, signal handlers, GC finalizers) *)
+let enqueue_thread t k x =
+  Lf_queue.push t.run_q (Thread (k, x));
+  if Atomic.get t.need_wakeup then wakeup t
+
+(* Safe to call from anywhere (other systhreads, domains, signal handlers, GC finalizers) *)
+let enqueue_failed_thread t k ex =
+  Lf_queue.push t.run_q (Failed_thread (k, ex));
+  if Atomic.get t.need_wakeup then wakeup t
+
+(* Can only be called from our own domain, so no need to check for wakeup. *)
+let enqueue_at_head t k =
+  Lf_queue.push_head t.run_q (Thread (k, ()))
+
+let get_waiters t fd =
+  match Hashtbl.find_opt t.fd_map fd with
+  | Some x -> x
+  | None ->
+    let x = { read = Lwt_dllist.create (); write = Lwt_dllist.create () } in
+    Hashtbl.add t.fd_map fd x;
+    x
+
+(* The OS told us that the event pipe is ready. Remove events. *)
+let clear_event_fd t =
+  let buf = Bytes.create 8 in   (* Read up to 8 events at a time *)
+  let got = Unix.read t.eventfd_r buf 0 (Bytes.length buf) in
+  assert (got > 0)
+
+(* Update [t.poll]'s entry for [fd] to match [waiters]. *)
+let update t waiters fd =
+  let fdi = Iomux.Util.fd_of_unix fd in
+  let flags =
+    match not (Lwt_dllist.is_empty waiters.read),
+          not (Lwt_dllist.is_empty waiters.write) with
+    | false, false -> Poll.Flags.empty
+    | true, false -> Poll.Flags.pollin
+    | false, true -> Poll.Flags.pollout
+    | true, true -> Poll.Flags.(pollin + pollout)
+  in
+  if flags = Poll.Flags.empty then (
+    Poll.invalidate_index t.poll fdi;
+    Hashtbl.remove t.fd_map fd
+  ) else (
+    Poll.set_index t.poll fdi fd flags;
+    if fdi > t.poll_maxi then
+      t.poll_maxi <- fdi
+  )
+
+let resume t node =
+  t.active_ops <- t.active_ops - 1;
+  let k : unit Suspended.t = Lwt_dllist.get node in
```

Comment:
Should we also just vendor in `Lwt_dllist` and give it a reasonable module name that is independent of Lwt?  I made a mistake of preserving the prefix when I originally pulled it out, and it's a little confusing to see a reference to Lwt at the heart of Eio when there is in fact none!

Reply:
Perhaps we should rename the `lwt_dllist` library itself (and provide `Lwt_dllist` as a compatibility alias)?

If we're going to change it in Eio, it should be a separate issue because we already depend on it elsewhere.

Reply:
I'd leave it as a Homage to Lwt :dancers:

Reply:
I'd live it as a Homage to Lwt :dancers: